### PR TITLE
Stdio work for #69

### PR DESCRIPTION
**Pull request by jrieken**
This PR adds support for standard in/out (in addition to http) to Omnisharp. Things are in ```Omnisharp.Stdio``` which is a console app reading from ```Console.In``` and writing to ```Console.Out```. Messages are JSON and conform to the v8 debugger protocol envelopes (https://code.google.com/p/chromedevtools/wiki/ChromeDevToolsProtocol). The implementation build on top of the ```Post```-MVC-attributes and reuses the Controllers defined in ```Omnisharp```. Log messages are wrapped into event-messages.

stdio

### CODE REVIEW

**Comment by nosami** [comment]
Have you got this code hooked up to the editor? 

Interested in knowing how it compares as far as performance goes. I'd expect it to be faster, but is it noticeable?

**Comment by filipw** [comment]
rather than manually reflecting over types you should use MVC infrastructure - in this case, `ActionDescriptors` (https://github.com/aspnet/Mvc/blob/c936ae80ca6b3df61e54e204444e088f5ced8051/src/Microsoft.AspNet.Mvc.Core/DefaultActionDescriptorsCollectionProvider.cs#L34)

This ensures that in the future, if any MVC components are added or customized the actions are still picked up correctly here

**Comment by jrieken** [comment]
@nosami - didn't compare perf yet. don't expect a big difference. the biggest win of this work is that it will enable 2-way communication without polling and assuming we factor out http into ```Omnisharp.Http``` less dependencies - faster startup, smaller bundle

**Comment by mat-mcloughlin** [comment]
I nominate @Yantrio to try adding it to brackets :)

**Comment by jrieken** [comment]
@filipw - You have a point - I will take a look at how to use it outside a normal MVC app.

**Comment by nosami** [comment]
Like @filipw, I'm not too keen on having the list of controllers here https://github.com/OmniSharp/omnisharp-roslyn/pull/86/files#diff-6dcf536ed4debf80eb70667622c41dd8R19 

Seems like we either need to solve that somehow, or have a test to ensure that all controller actions are reachable via stdio. Otherwise, we're bound to forget to update that list at some point!

Also, I'm not too keen on having 2 Program.cs files, 2 StartUp.cs files etc. They'd need to be kept in sync. Maybe we could address that in a subsequent PR?

File: src/OmniSharp.Stdio/Logging/StdioLoggerProvider.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+using System;
+using Microsoft.Framework.Logging;
+
+namespace OmniSharp.Stdio.Logging
+{
+    internal class StdioLoggerProvider : ILoggerProvider
+    {
+
```
Comment (davidfowl) [inline src/OmniSharp.Stdio/Logging/StdioLoggerProvider.cs]: Extra space!

File: src/OmniSharp.Stdio/Program.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                if (arg == "-s")
+                {
+                    enumerator.MoveNext();
+                    applicationRoot = Path.GetFullPath((string)enumerator.Current);
+                }
+                else if (arg == "-v")
+                {
+                    logLevel = LogLevel.Verbose;
+                }
+                else if (arg == "--hostPID")
```
Comment (davidfowl) [inline src/OmniSharp.Stdio/Program.cs]: Useless for stdio

File: src/OmniSharp.Stdio/Program.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            });
+
+            Console.WriteLine(new EventPacket()
+            {
+                Event = "started"
+            });
+
+            while (!appShutdownService.ShutdownRequested.IsCancellationRequested)
+            {
+                var line = Console.ReadLine();
```
Comment (davidfowl) [inline src/OmniSharp.Stdio/Program.cs]: Do what the other program.main does and wait on ctrl+c

File: src/OmniSharp.Stdio/Protocol/Packet.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+        public int Seq { get; set; }
+
+        public string Type { get { return _type; } }
+
+        public override string ToString() {
+            return JsonConvert.SerializeObject(this, new JsonSerializerSettings() { ContractResolver = LowercaseContractResolver.Instance });
+        }
+
+        private class LowercaseContractResolver : DefaultContractResolver
```
Comment (davidfowl) [inline src/OmniSharp.Stdio/Protocol/Packet.cs]: Nooo, use the CamelCase one

File: src/OmniSharp.Stdio/Protocol/RequestPacket.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace OmniSharp.Stdio.Protocol
+{
+    public class RequestPacket : Packet
+    {
+        private readonly JObject obj;
```
Comment (davidfowl) [inline src/OmniSharp.Stdio/Protocol/RequestPacket.cs]: _obj

File: src/OmniSharp.Stdio/project.json [hunk 6]
```diff
@@ -0,0 +1,20 @@
+﻿{
+    "version": "1.0.0-*",
+    "dependencies": {
+        "OmniSharp": "1.0.0-*",
+        "Microsoft.Framework.Runtime": "1.0.0-beta2",
+        "Microsoft.AspNet.Hosting": "1.0.0-beta2",
+        "Newtonsoft.Json": "6.0.8"
+    },
+    "commands": { 
+        "run" : "run"
```
Comment (davidfowl) [inline src/OmniSharp.Stdio/project.json]: ugh, can we change this to something else. Like:

```JSON
"commands": {
    "OmniSharp.Stdio": "OmniSharp.Stdio"
}
```

**Comment by jrieken** [comment]
Pulling back this PR - I have a solution for this which is sooo much cooler ;-) 

It does pick up @davidfowl crazy idea of asp on stdio. Basically implementing a OWIN server that reads/writes from/to the console. That way we can keep one executable and keep using all fancy ASP and MVC features.

**Comment by filipw** [comment]
sounds awesome!

**Comment by nosami** [comment]
The crazier the better, I reckon!

**Comment by jrieken** [comment]
work is happening here: https://github.com/OmniSharp/omnisharp-roslyn/tree/stdio2. almost done, lacks tests. off for a few days :ski:

**Comment by nosami** [comment]
Heh! I've been liking at it already :) looks like a big improvement to me

**Comment by jrieken** [comment]
new work is in PR #90

**Pull request closed without merging**